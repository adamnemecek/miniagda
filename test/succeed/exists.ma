-- 2010-03-28 Exists and Bracket via parametric function types
-- 2012-01-22 parameters gone from constructors

data Sigma (A : Set)(B : A -> Set) : Set
{ pair : (fst : A) -> (snd : B fst) -> Sigma A B
}

data Subset (A : Set)(B : A -> Set) : Set
{ put : (get : A) -> [prf : B get] -> Subset A B
}

data Exists (A : Set)(B : A -> Set) : Set
{ exI : [a : A] -> (prop : B a) -> Exists A B
}

fun exE : [A : Set] -> [B : A -> Set] -> [C : Set] ->
      Exists A B -> ([a : A] -> B a -> C) -> C
{ exE A B C (exI a b) k = k a b
}

data Bracket (A : Set) : Set
{ bI : [a : A] -> Bracket A
}

fun bE : [A : Set] -> [C : Set] -> Bracket A -> ([A] -> C) -> C
{ bE A C (bI a) k = k a
}

