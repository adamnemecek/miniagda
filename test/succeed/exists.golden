MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "exists.ma" ---
--- scope checking ---
--- type checking ---
type  Sigma : ^(A : Set) -> ^(B : A -> Set) -> Set
term  Sigma.pair : .[A : Set] -> .[B : A -> Set] -> ^(fst : A) -> ^(snd : B fst) -> < Sigma.pair fst snd : Sigma A B >
term  fst : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> A
{ fst [A] [B] (Sigma.pair #fst #snd) = #fst
}
term  snd : .[A : Set] -> .[B : A -> Set] -> (pair : Sigma A B) -> B (fst [A] [B] pair)
{ snd [A] [B] (Sigma.pair #fst #snd) = #snd
}
type  Subset : ^(A : Set) -> ^(B : A -> Set) -> Set
term  Subset.put : .[A : Set] -> .[B : A -> Set] -> ^(get : A) -> .[prf : B get] -> < Subset.put get prf : Subset A B >
type  Exists : ^(A : Set) -> ^(B : A -> Set) -> Set
term  Exists.exI : .[A : Set] -> .[B : A -> Set] -> .[a : A] -> ^(prop : B a) -> < Exists.exI a prop : Exists A B >
term  exE : .[A : Set] -> .[B : A -> Set] -> .[C : Set] -> Exists A B -> (.[a : A] -> B a -> C) -> C
{ exE [A] [B] [C] (Exists.exI [a] b) k = k [a] b
}
type  Bracket : ^(A : Set) -> Set
term  Bracket.bI : .[A : Set] -> .[a : A] -> < Bracket.bI a : Bracket A >
term  bE : .[A : Set] -> .[C : Set] -> Bracket A -> (.[A] -> C) -> C
{ bE [A] [C] (Bracket.bI [a]) k = k [a]
}
--- evaluating ---
--- closing "exists.ma" ---
